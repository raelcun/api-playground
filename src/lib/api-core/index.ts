export * from '../../modules/error'
export * from './response'
