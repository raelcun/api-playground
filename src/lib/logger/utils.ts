import { either as E, taskEither as TE } from 'fp-ts'

import { Err } from '@lib/error'

import { Logger, LogMethods } from './types'

export const logErrorsE = <L extends Err<string>>(logger: Logger, method: LogMethods = 'trace') =>
  E.mapLeft<L, L>(e => {
    logger.log({ level: method, payload: e })
    return e
  })

export const logErrorsTE = <L extends Err<string>>(logger: Logger, method: LogMethods = 'trace') =>
  TE.mapLeft<L, L>(e => {
    logger.log({ level: method, payload: e })
    return e
  })
