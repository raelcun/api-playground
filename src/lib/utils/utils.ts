import { array as A, either as E, io as IO, taskEither as TE } from 'fp-ts'
import { pipe } from 'fp-ts/lib/pipeable'
import * as t from 'io-ts'
import { reporter } from 'io-ts-reporters'
import { Context, Middleware } from 'koa'

import { Err } from '@lib/error'

export const tryCatchTE: typeof TE.tryCatch = (f, onRejected) =>
  TE.tryCatch(() => {
    try {
      return f()
    } catch (err) {
      return Promise.reject(err)
    }
  }, onRejected)
tryCatchTE.call
export const createVoidTE = <L>() => TE.rightTask<L, void>(async () => {})

export const filterObjectKeys = <T extends Record<string, unknown>, K extends keyof T>(
  o: T,
  keys: K[],
): Pick<T, K> => {
  const res: any = {}

  keys.forEach(key => {
    res[key] = o[key]
  })

  return res
}

export const mapErrorCode = <T extends string>(newErrorCode: T) =>
  E.mapLeft((originalError: Err<string>) => ({
    ...originalError,
    code: newErrorCode,
    subcode: originalError.code,
  }))

export const decode = <T>(
  t: t.Type<T, unknown>,
  v: unknown,
): E.Either<Err<'VALIDATION_ERROR'>, T> =>
  pipe(
    t.decode(v),
    E.mapLeft(errors => ({
      code: 'VALIDATION_ERROR',
      message: reporter(E.left(errors)).join('. '),
    })),
  )

export const ioSequence = A.array.sequence(IO.io)

export const DateFromString = new t.Type<Date, string, unknown>(
  'DateFromString',
  (u): u is Date => u instanceof Date,
  (u, c) =>
    pipe(
      t.union([t.string, t.number]).validate(u, c),
      E.chain(s => {
        const d = new Date(s)
        return isNaN(d.getTime()) ? t.failure(u, c) : t.success(d)
      }),
    ),
  a => a.toISOString(),
)

export const createMiddlewareTE = <T>(
  f: (ctx: Context) => TE.TaskEither<unknown, unknown>,
): Middleware<T> => async (ctx, next) => {
  await pipe(
    f(ctx),
    TE.chain(() => TE.rightTask(next)),
  )()
}

export const createMiddlewareE = <T>(
  f: (ctx: Context) => E.Either<unknown, unknown>,
): Middleware<T> => createMiddlewareTE(ctx => TE.fromEither(f(ctx)))
