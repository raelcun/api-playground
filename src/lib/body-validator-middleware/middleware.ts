import { either as E } from 'fp-ts'
import HttpStatus from 'http-status-codes'
import * as t from 'io-ts'
import { Middleware as KoaMiddleware } from 'koa'

import { validateBody } from '@lib/body-validator'
import { LoggerFactory } from '@lib/logger'
import { createErrorResponse, KoaContext, Middleware } from '@modules/api-core'

export const withValidatedBody = (createLogger: LoggerFactory) => <T>(type: t.Type<T, unknown>) => (
  middleware: Middleware<T>,
): KoaMiddleware => async (ctx, next) => {
  const result = validateBody(createLogger)(type)(ctx.request.body)

  if (E.isLeft(result)) {
    ctx.status = HttpStatus.BAD_REQUEST
    ctx.body = createErrorResponse(result.left)
    return
  }

  await middleware(ctx as KoaContext<T>, next)
}
