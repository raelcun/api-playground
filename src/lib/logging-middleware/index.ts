import { getSystemLogger } from '@lib/logger'
import { createMiddleware } from '@modules/logging-middleware'

export const middleware = createMiddleware(getSystemLogger)
