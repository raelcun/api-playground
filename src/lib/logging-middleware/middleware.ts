import { Middleware } from 'koa'
import { v4 as uuid } from 'uuid'

import { createTimingMessageSegment, getMessageFromContext, LoggerProvider } from '@lib/logger'
import { getLogger2 } from '@modules/logger'

export const createMiddleware = (getLogger: LoggerProvider): Middleware => async (ctx, next) => {
  const startTime = new Date()
  const startNanos = process.hrtime.bigint()

  ctx.state.logger = getLogger2().pinValues({
    traceId: uuid(),
    ...getMessageFromContext(ctx, { request: true, response: false }),
  })

  let err: any = undefined
  try {
    await next()
  } catch (e) {
    err = e
  }

  getLogger().info({
    ...getMessageFromContext(ctx),
    timing: createTimingMessageSegment(startNanos, startTime, 'RESPONSE_TIMING'),
  })

  if (err !== undefined) throw err
}
