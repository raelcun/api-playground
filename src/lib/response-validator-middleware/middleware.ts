import { either as E } from 'fp-ts'
import { Lazy } from 'fp-ts/lib/function'
import { pipe } from 'fp-ts/lib/pipeable'
import HttpStatus from 'http-status-codes'
import * as t from 'io-ts'
import { Middleware } from 'koa'

import { Err } from '@lib/error'
import { LoggerFactory } from '@lib/logger'
import { decode, logErrors, mapErrorCode } from '@lib/utils'

const validateResponseInternal = (createLogger: LoggerFactory) => <T>(type: t.Type<T, unknown>) => (
  response: unknown,
): E.Either<Err<'RESPONSE_VALIDATION_ERROR'>, T> =>
  pipe(
    decode(type, response),
    mapErrorCode('RESPONSE_VALIDATION_ERROR'),
    logErrors(createLogger(), 'warn'),
  )

export const validateResponseMiddleware = (
  createLogger: LoggerFactory,
  shouldError: Lazy<boolean>,
) => <T>(type: t.Type<T, unknown>): Middleware => async (ctx, next) => {
  pipe(
    validateResponseInternal(createLogger)(type)(ctx.body),
    E.fold(
      () => {
        if (shouldError()) {
          ctx.status = HttpStatus.INTERNAL_SERVER_ERROR
        } else {
          next()
        }
      },
      () => {
        next()
      },
    ),
  )
}
