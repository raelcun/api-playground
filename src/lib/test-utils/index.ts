import * as HttpStatus from 'http-status-codes'
import { Context } from 'koa'

import { Logger } from '@lib/logger'
import { install, InstalledClock } from '@sinonjs/fake-timers'

export const createMockLogger = (): Logger => ({
  info: jest.fn(),
  fatal: jest.fn(),
  trace: jest.fn(),
  error: jest.fn(),
  debug: jest.fn(),
  warn: jest.fn(),
  child: jest.fn(),
})

export const testUnauthorized = (mockContext: Context, mockNext: jest.Mock<Promise<void>, []>) => {
  test('should set status to 401', () => {
    expect(mockContext.status).toEqual(HttpStatus.UNAUTHORIZED)
  })

  test('should not invoke next', () => {
    expect(mockNext).not.toHaveBeenCalled()
  })
}

export const testAuthorized = (mockContext: Context, mockNext: jest.Mock<Promise<void>, []>) => {
  test('should not set status to 401', () => {
    expect(mockContext.status).not.toEqual(HttpStatus.UNAUTHORIZED)
  })

  test('should invoke next once', () => {
    expect(mockNext).toHaveBeenCalledTimes(1)
  })
}

/* istanbul ignore next */
export const installStaticClock = () => {
  let clock: InstalledClock

  beforeAll(() => {
    clock = install({ now: 946684800000 })
  })

  afterAll(() => {
    clock.uninstall()
  })

  return () => clock
}
