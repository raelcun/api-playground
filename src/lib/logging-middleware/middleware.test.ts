import { Lazy } from 'fp-ts/lib/function'
import { Next } from 'koa'

import { createMockLogger, installStaticClock } from '@lib/test-utils'
import { createKoaContext, createMockNext } from '@modules/utils'

import { createMiddleware } from './middleware'

const eatError = async <T>(fn: Lazy<Promise<T>>) => {
  try {
    await fn()
  } catch {}
}

const createMockMiddleware = () => {
  const mockLogger = createMockLogger()
  const mockInfo = mockLogger.info as ReturnType<typeof jest.fn>
  const mockContext = createKoaContext({ url: 'MOCK_URL', method: 'GET' })
  const middleware = createMiddleware(() => mockLogger)

  return {
    mockInfo,
    mockContext,
    callMiddleware: (next: Next) => middleware(mockContext, next),
  }
}

describe('logRequestMiddleware', () => {
  const getClock = installStaticClock()

  test('log should match snapshot when next throws', async () => {
    const { callMiddleware, mockInfo, mockContext } = createMockMiddleware()
    const mockNext = createMockNext(() => {
      getClock().tick(1)
      mockContext.status = 500
      throw new Error()
    })
    await eatError(callMiddleware(mockNext))
    expect(mockInfo.mock.calls[0][0]).toMatchSnapshot()
  })

  test('log should match snapshot when next succeeds', async () => {
    const { callMiddleware, mockInfo, mockContext } = createMockMiddleware()
    const mockNext = createMockNext(() => {
      getClock().tick(1)
      mockContext.status = 200
      throw new Error()
    })
    await eatError(() => callMiddleware(mockNext))
    expect(mockInfo.mock.calls[0][0]).toMatchSnapshot()
  })

  test('should pass through error', async () => {
    expect.assertions(1)

    const { callMiddleware } = createMockMiddleware()
    const mockNext = createMockNext(() => {
      throw new Error('foobar')
    })

    try {
      await callMiddleware(mockNext)
    } catch (e) {
      expect(e.message).toEqual('foobar')
    }
  })
})
