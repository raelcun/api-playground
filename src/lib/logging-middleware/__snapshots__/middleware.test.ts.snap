// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logRequestMiddleware log should match snapshot when next succeeds 1`] = `
Object {
  "request": Object {
    "headers": Object {
      "host": "test.com",
    },
    "method": "GET",
    "url": "http://test.com/MOCK_URL",
  },
  "response": Object {
    "body": undefined,
    "headers": Object {},
    "status": 200,
  },
  "timing": Object {
    "duration": 1,
    "end": "2000-01-01T00:00:00.002Z",
    "start": "2000-01-01T00:00:00.001Z",
    "type": "RESPONSE_TIMING",
  },
}
`;

exports[`logRequestMiddleware log should match snapshot when next throws 1`] = `
Object {
  "request": Object {
    "headers": Object {
      "host": "test.com",
    },
    "method": "GET",
    "url": "http://test.com/MOCK_URL",
  },
  "response": Object {
    "body": undefined,
    "headers": Object {},
    "status": 500,
  },
  "timing": Object {
    "duration": 1,
    "end": "2000-01-01T00:00:00.001Z",
    "start": "2000-01-01T00:00:00.000Z",
    "type": "RESPONSE_TIMING",
  },
}
`;
